runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			/* Every consecutivePGCToChangeEden number of pgc's, re-evaluate eden size, and move towards it */
+			moveTowardRecommendedEden(env, 0.25);
+		}
+	} else if (0 == _pgcCountSinceGMPEnd % 3) {
+		/*
+		 * On every third pgc, make an adjustment to eden size based on observed pgc values. 
+		 * Waiting for every third PGC, allows some statistics (which are being averaged) to settle to their new true averages
+ 		 */
+
+		intptr_t edenRegionChange = 0;
+		intptr_t edenChangeMagnitude = (intptr_t)ceil((0.1 * getIdealEdenSizeInBytes(env)) / _regionManager->getRegionSize());
+
+		double hybridEdenOverhead = calculateHybridEdenOverhead(env, _historicalPartialGCTime, _partialGcOverhead);
+		
+		/* 
+		 * Aim to get hybrid PGC overhead between extensions->dnssExpectedTimeRatioMinimum and extensions->dnssExpectedTimeRatioMaximum 
+		 * by increasing or decreasing eden by 10% 
+		 */
+		if (_extensions->dnssExpectedTimeRatioMinimum._valueSpecified > hybridEdenOverhead ) {
+			/* Shrink eden a bit */
+			edenRegionChange = edenChangeMagnitude * -1;
+		} else if (_extensions->dnssExpectedTimeRatioMaximum._valueSpecified < hybridEdenOverhead) {
+			/* Expand eden a bit */
+			edenRegionChange = edenChangeMagnitude;
+		}
+
+		_edenSizeFactor += edenRegionChange;	
+	}
+}
+
+double 
+MM_SchedulingDelegate::mapPgcTimeToPgcOverhead(MM_EnvironmentVLHGC *env, uintptr_t partialGcTimeMs) {
+	
+	/* Convert expectedTimeRatioMinimum/Maximum to 0-100 based for this formula */
+	double xminpct = _extensions->dnssExpectedTimeRatioMinimum._valueSpecified * 100;
+	double xmaxpct = _extensions->dnssExpectedTimeRatioMaximum._valueSpecified * 100;
+	double xmaxt = (double)_extensions->tarokTargetMaxPauseTime;
+
+	double overhead;
+
+	if (heapIsFullyExpanded(env)) {
+		/* 
+		 * Eden size is being driven by heuristic which is trying to MINIMIZE hybrid overead. A low avg pgc time, is more desirable than high avg time,
+		 * So the overhead logic needs to map a low avg pgc time, to a low overhead value (aka, a "better"/more desirable value)
+		 * Ex 20ms -> 5% (good/desirable), 1000ms -> 80% (bad/undesirable/eden should probably shrink)
+		 */
+		double midpointPct = (xmaxpct + xminpct)/2.0;
+		if (partialGcTimeMs <= xmaxt) {
+			/* Once the pgc time is at, or below the max pgc time, there is no "benefit" from shrinking it further, since we are already satisfying tarokTargetMaxPauseTime */
+			overhead = midpointPct;
+		} else {
+			/* 
+			 * If pgc time is above the max pgc time, map high PGC time values as very very high overhead, in efforts to bring the PGC time down to tarokTargetMaxPauseTime
+			 * If pgc time is only slightly above tarokTargetMaxPauseTime, then there is only a very small overhead penalty, 
+			 * wheras being 2x higher than the target pause time leads to a significantly bigger penalty  
+			 */
+			double overheadCurve = pow(1.03, ((double)partialGcTimeMs - xmaxt)) + midpointPct - 1;
```

Comment:
This pow() will have very different behavior for potentially small xmaxt (1-10ms) vs large ones (100ms -1s). The point is, 2x violation should lead to the same overheadCurve value, regardless what the threshold is.

I'd like to see some normalization. Something like this. 

( base ^ ( partialGcTimeMs - xmaxt)/xmaxt  ) x midpointPct  + midpointPct 

'base' could be very different from what you used.

Reply:
That makes sense, I'll fix that up